### PR TITLE
qa: add iSCSI gateway smoke test

### PR DESCRIPTION
**Pull request by smithfarm**
Note: merge https://github.com/SUSE/ceph/pull/136 when this is merged.

### CODE REVIEW

**Comment by smithfarm** [comment]
Test result is here: http://147.135.133.225:8081/ubuntu-2017-07-17_21:04:40-deepsea:basic:igw-ses5---basic-openstack/

**Comment by smithfarm** [comment]
The test case appears to go fine until "iscsiadm -m discovery", which fails like this:

```
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    + iscsiadm -m discovery -t st -p target137074024206.teuthology
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: cannot make connection to 137.74.24.206: Connection refused
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: cannot make connection to 137.74.24.206: Connection refused
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: cannot make connection to 137.74.24.206: Connection refused
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: cannot make connection to 137.74.24.206: Connection refused
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: cannot make connection to 137.74.24.206: Connection refused
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: cannot make connection to 137.74.24.206: Connection refused
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: connection login retries (reopen_max) 5 exceeded
2017-07-17T21:15:15.518 INFO:teuthology.orchestra.run.target137074024206.stdout:    iscsiadm: Could not perform SendTargets discovery: encountered connection failure
```

**Comment by smithfarm** [comment]
I'll be grateful if anyone can suggest how to debug this further.

**Comment by ddiss** [comment]
It would appear that the iSCSI target node at $IGWNODE isn't accepting connections. Would it be possible to obtain lrbd logs?

**Comment by smithfarm** [comment]
@ddiss Lots of lrbd-related messages in `misc.log.gz` at http://147.135.133.225/ubuntu-2017-07-17_21:04:40-deepsea:basic:igw-ses5---basic-openstack/130/remote/target137074024206/syslog/

If lrbd has a specific log file, please tell me the name and I'll add it to the list of log files that teuthology collects. Also, how can I turn on debug-level lrbd logging?

**Comment by smithfarm** [comment]
Also, which command(s) could I run server-side (i.e. on the igw node) to assert that the Target IQN (sorry if that's the wrong term, but hopefully you know what i mean) is being properly advertised?

**Comment by ddiss** [comment]
@smithfarm the easiest way to confirm LIO setup would be by dumping the contents of the directory tree under /sys/kernel/config/target/ . You could also use "netstat -l" to check whether the kernel is listening on iSCSI port 3260.

**Comment by smithfarm** [comment]
@ddiss Thanks, I added that. Now, if you could take a look at this log: http://147.135.133.225/ubuntu-2017-07-19_22:27:30-deepsea:basic:igw-ses5---basic-openstack/144/teuthology.log

Note: text search for "igw info script" will take you directly to the interesting bit. (Spoiler: 3260 does not appear in netstat)

**Comment by ddiss** [comment]
The iSCSI target hasn't been configured with a network portal (which explains the lack of a port 3260 listener) or a logical unit:

2017-07-19T22:37:58.420 INFO:teuthology.orchestra.run.target149202164181.stdout:    /sys/kernel/config/target/iscsi/iqn.2003-01.org.linux-iscsi.target149202164159.x8664:sn.c30a5c43a9d6/tpgt_1/lun:
2017-07-19T22:37:58.420 INFO:teuthology.orchestra.run.target149202164181.stdout:    total 0
2017-07-19T22:37:58.420 INFO:teuthology.orchestra.run.target149202164181.stdout:
2017-07-19T22:37:58.478 INFO:teuthology.orchestra.run.target149202164181.stderr:[DEBUG   ] jid 20170719223758103822 found all minions set(['target149202164159.teuthology'])
2017-07-19T22:37:58.516 INFO:teuthology.orchestra.run.target149202164181.stdout:    /sys/kernel/config/target/iscsi/iqn.2003-01.org.linux-iscsi.target149202164159.x8664:sn.c30a5c43a9d6/tpgt_1/np:
2017-07-19T22:37:58.517 INFO:teuthology.orchestra.run.target149202164181.stdout:    total 0

Please provide the lrbd configuration.

File: qa/common/common.sh [hunk 1]
```diff
@@ -267,6 +267,26 @@ EOF
...
+  # peek at igw information on the igw node
+  #
+  local TESTSCRIPT=/tmp/igw_info.sh
+  local IGWNODE=$(_first_x_node igw)
+  cat << 'EOF' > $TESTSCRIPT
+set -ex
+trap 'echo "Result: NOT_OK"' ERR
+echo "igw info script running as $(whoami) on $(hostname --fqdn)"
+rpm -q lrbd || true
+lrbd --output || true
```
Comment (smithfarm) [inline qa/common/common.sh]: added `lrbd --output`, test now running at http://147.135.133.225:8081/ubuntu-2017-07-20_09:03:38-deepsea:basic:igw-ses5---basic-openstack/

**Comment by smithfarm** [comment]
Output of lrbd command with `--debug` added: http://pastebin.nue.suse.com/23114/src

**Comment by smithfarm** [comment]
Thanks to @ddiss for finding a workaround: replace `"host": "target149202172019",` with`"host": "target149202172019.teuthology",` in `/tmp/lrbd.conf` and then re-run `lrbd`.

**Comment by smithfarm** [comment]
Opened https://bugzilla.suse.com/show_bug.cgi?id=1049669 to track the lrbd issue.

**Comment by smithfarm** [comment]
@ddiss Progress, but still no cigar: http://pastebin.nue.suse.com/23182/src

**Comment by smithfarm** [comment]
Finally, the test passes: http://pastebin.nue.suse.com/23186/src

I'd say this PR is now ready to merge.

**Comment by smithfarm** [comment]
The new instrumentation yields the following info: http://pastebin.nue.suse.com/23230/src

I guess we could use `/dev/disk/by-path/ip-213.32.78.223:3260-iscsi-iqn.2016-11.org.linux-iscsi.igw.x86:sn.demo-lun-0` but how to determine that programmatically?

**Comment by ddiss** [comment]
> I guess we could use /dev/disk/by-path/ip-213.32.78.223:3260-iscsi-iqn.2016-11.org.linux-iscsi.igw.x86:sn.demo-lun-0 but how to determine that programmatically?

But that only corresponds to one of the paths, rather than the multipath device, right? For the multipath device, there should be a by-scsi-id or similar that links to the dm-X multipath device path.

**Comment by smithfarm** [comment]
@ddiss Yes, there are several links to the dm-0 device:

```
2017-07-21T10:16:03.109 INFO:teuthology.orchestra.run.target213032078237.stdout:    + ls -lR /dev/mapper
2017-07-21T10:16:03.109 INFO:teuthology.orchestra.run.target213032078237.stdout:    /dev/mapper:
2017-07-21T10:16:03.109 INFO:teuthology.orchestra.run.target213032078237.stdout:    total 0
2017-07-21T10:16:03.109 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root       7 Jul 21 10:15 36001405f8ad312273b232edbe9f9447f -> ../dm-0
2017-07-21T10:16:03.109 INFO:teuthology.orchestra.run.target213032078237.stdout:    crw------- 1 root root 10, 236 Jul 21 10:15 control
2017-07-21T10:16:03.109 INFO:teuthology.orchestra.run.target213032078237.stdout:    + ls -l /dev/disk/by-path
2017-07-21T10:16:03.110 INFO:teuthology.orchestra.run.target213032078237.stdout:    total 0
2017-07-21T10:16:03.110 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root 9 Jul 21 10:15 ip-213.32.78.223:3260-iscsi-iqn.2016-11.org.linux-iscsi.igw.x86:sn.demo-lun-0 -> ../../sda
2017-07-21T10:16:03.110 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root 9 Jul 21 10:15 scsi-0:0:0:0 -> ../../sda
2017-07-21T10:16:03.110 INFO:teuthology.orchestra.run.target213032078237.stdout:    + ls -l /dev/disk/by-id /dev/disk/by-parttypeuuid /dev/disk/by-partuuid /dev/disk/by-uuid
2017-07-21T10:16:03.110 INFO:teuthology.orchestra.run.target213032078237.stdout:    /dev/disk/by-id:
2017-07-21T10:16:03.110 INFO:teuthology.orchestra.run.target213032078237.stdout:    total 0
2017-07-21T10:16:03.110 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root 10 Jul 21 10:15 dm-name-36001405f8ad312273b232edbe9f9447f -> ../../dm-0
2017-07-21T10:16:03.111 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root 10 Jul 21 10:15 dm-uuid-mpath-36001405f8ad312273b232edbe9f9447f -> ../../dm-0
2017-07-21T10:16:03.111 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root  9 Jul 21 10:15 scsi-1SUSE_RBD:f8ad3122-73b2-32ed-be9f-9447f9830584 -> ../../sda
2017-07-21T10:16:03.111 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root 10 Jul 21 10:15 scsi-36001405f8ad312273b232edbe9f9447f -> ../../dm-0
2017-07-21T10:16:03.111 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root  9 Jul 21 10:15 scsi-SSUSE_RBD_f8ad3122-73b2-32ed-be9f-9447f9830584 -> ../../sda
2017-07-21T10:16:03.111 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root  9 Jul 21 10:09 virtio-6006e8a1-47f4-47d3-a -> ../../vdc
2017-07-21T10:16:03.111 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root  9 Jul 21 10:09 virtio-634962f6-fa39-42ea-b -> ../../vdd
2017-07-21T10:16:03.112 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root  9 Jul 21 10:08 virtio-8cfbe00f-1912-4fab-a -> ../../vdb
2017-07-21T10:16:03.112 INFO:teuthology.orchestra.run.target213032078237.stdout:    lrwxrwxrwx 1 root root 10 Jul 21 10:15 wwn-0x6001405f8ad312273b232edbe9f9447f -> ../../dm-0
2017-07-21T10:16:03.112 INFO:teuthology.orchestra.run.target213032078237.stdout:
```

**Pull request merged**